File: errand/src/lib/orchestrator/stream.ts
import type { AuditEvent, AuditSink } from "@/lib/contracts";

/**
 * Streaming transport for orchestrator events.
 *
 * Client <-> server is streaming-only, NO client polling. The orchestrator
 * records an AuditEvent at every meaningful step (context.loaded, cart.built,
 * approval.granted, payment.credential, checkout.completed, mail.confirmation,
 * ...). A StreamingAuditSink pushes each event to the client the instant it
 * happens, so the UI reflects real-time progress without ever polling.
 *
 * Server <-> third-party polling (Prava payment-result, AgentMail inbox) is
 * internal to the brokers and hidden from the client; its results surface here
 * as streamed events.
 */

export type EmitFn = (event: AuditEvent) => void;

/** AuditSink that forwards every event to an emit callback (and keeps a copy). */
export class StreamingAuditSink implements AuditSink {
  private events: AuditEvent[] = [];
  constructor(private readonly emit: EmitFn) {}
  record(event: AuditEvent): void {
    this.events.push(event);
    this.emit(event);
  }
  all(): AuditEvent[] {
    return [...this.events];
  }
}

/** SSE frame for one event. Use in a Next.js Route Handler ReadableStream. */
export function sseFrame(event: AuditEvent): string {
  return `event: ${event.step}\ndata: ${JSON.stringify(event)}\n\n`;
}

/**
 * Build a streaming Response body for a Route Handler. `run` receives an `emit`
 * it calls (via a StreamingAuditSink) for each step; the returned ReadableStream
 * flushes each event to the client immediately, then closes.
 *
 * Example (app/api/errand/route.ts):
 *   return sseResponse(async (emit) => {
 *     const audit = new StreamingAuditSink(emit);
 *     await runErrand({ ...brokers, audit }, input);
 *   });
 */
export function sseResponse(
  run: (emit: EmitFn) => Promise<void>,
): Response {
  const encoder = new TextEncoder();
  const stream = new ReadableStream<Uint8Array>({
    async start(controller) {
      const emit: EmitFn = (event) => {
        controller.enqueue(encoder.encode(sseFrame(event)));
      };
      try {
        await run(emit);
        controller.enqueue(encoder.encode("event: done\ndata: {}\n\n"));
      } catch (err) {
        const message = err instanceof Error ? err.message : String(err);
        controller.enqueue(
          encoder.encode(`event: error\ndata: ${JSON.stringify({ message })}\n\n`),
        );
      } finally {
        controller.close();
      }
    },
  });
  return new Response(stream, {
    headers: {
      "Content-Type": "text/event-stream",
      "Cache-Control": "no-cache, no-transform",
      Connection: "keep-alive",
    },
  });
}

File: errand/backend/app/orchestrator/stream.py
"""Real-time SSE streaming. Client <-> server is streaming-only; the client
NEVER polls. Orchestrator events are pushed the instant they happen.
"""

from __future__ import annotations

import asyncio
import json
from typing import AsyncIterator

from app.contracts import AuditEvent


def sse_frame(event: str, data: dict) -> str:
    return f"event: {event}\ndata: {json.dumps(data)}\n\n"


class EventStream:
    """A bounded async queue that the orchestrator writes to and the SSE
    response drains. `emit` is passed to run_errand; `drain` yields frames."""

    def __init__(self) -> None:
        self._q: asyncio.Queue[tuple[str, dict] | None] = asyncio.Queue()

    async def emit(self, event: AuditEvent) -> None:
        await self._q.put((event.step, event.model_dump()))

    async def emit_raw(self, event: str, data: dict) -> None:
        await self._q.put((event, data))

    async def close(self) -> None:
        await self._q.put(None)

    async def drain(self) -> AsyncIterator[str]:
        while True:
            item = await self._q.get()
            if item is None:
                break
            event, data = item
            yield sse_frame(event, data)
